### PR TITLE
docs(SCHEMA): add SCHEMA.md to document HOCON schema

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

File: SCHEMA.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  NOTE: For generic override, see [below](#default_override_rule) for more info
+* `nullable`: set to `true` if this field is allowed to be `undefined`.
+  NOTE: there is no point setting it to `true` if fields has a default value.
+* `sensitive`: set to `true` if this field's value is sensitive so we will obfuscate the log
+  with `********` when logging.
+
+<a name="default_override_rule"></a>
+## Default environment variable override
+
+By defulat, a field except for array elements can be overriden by an environment variable if
```
Comment (k32) [inline SCHEMA.md:204]:
```suggestion
By default, a field except for array elements can be overriden by an environment variable if
```

File: SCHEMA.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
+# HOCON schema
+
+HOCON schema is a type-safe data validation framework inspired
+by [basho cuttlefish](https://github.com/basho/cuttlefish)
+
+## Types
+
+There are two high level kinds of data types defined in HOCON schema: `primitve` and `complex`.
+If we think of Erlang data structure as a 'tree', then `primitive` types are to define the
```
Comment (k32) [inline SCHEMA.md:9]: are to define -> denote ?

File: SCHEMA.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+'leaves' of the 'tree'. aka the terminal values. While the `complex` types are to define
```
Comment (k32) [inline SCHEMA.md:10]: Same here

File: SCHEMA.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+values which encloses either another `complex` or `primitive` values.
+
+### Primitive types
+
+Most of the `primitive` types are provided by (and also can be extended from) the [typerefl](https://github.com/k32/typerefl) libray.
+Typerefl is highly composible hence can be used to define complex types.
+However in HOCON schema we only use it to define `primitive` types.
+
+Here is a list of the `primitive` types for reference.
```
Comment (k32) [inline SCHEMA.md:19]:
```suggestion
Here is a list of the `primitive` types for reference:
```

File: SCHEMA.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+* enum: Enum is a list of Erlang atoms
+* singleton: singleton is an Erlang atom
+* integer: typerefl:integer()
+* string: typerefl:integer()
+
+And a extended primitive type example: `ip_port`
```
Comment (k32) [inline SCHEMA.md:26]:
```suggestion
And an extended primitive type example: `ip_port`
```

File: SCHEMA.md [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+```
+
+### Complex types
+
+HOCON schema supports 3 different complext types: `struct`, `array`, and `union`.
+NOTE: to make it easier for future extensions, it's recommended to use `hoconsc` module APIs to define schema.
+
+#### Structs
+
+NOTE: HOCON schema does not support non-struct too level data types. e.g. it is not allowed to
```
Comment (k32) [inline SCHEMA.md:44]: > non-struct too level data types

Typo?

File: SCHEMA.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+define a root level schema with just a `integer()` type.
+
+Structs consist of data fields, which can be defiend using `hocon_schema` behaviour callbacks.
```
Comment (k32) [inline SCHEMA.md:47]:
```suggestion
Structs consist of data fields, which can be defined using `hocon_schema` behaviour callbacks.
```

File: SCHEMA.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+fields("foo") -> [{"field1", typerefl:integer()}].
+```
+
+In this case, the schema for use in `hocon_schema` APIs is the module name. There is another way to
+define a struct as a Erlang `map()`, so we do not have to implement the behaviour callbacks
+(this is however mostly for test cases):
+
+```
+#{structs => ["foo"], %% 'exported' root names
+  'fields => #{"foo" => [{"field1", typerefl:integer()}]}
```
Comment (k32) [inline SCHEMA.md:66]:
```suggestion
  , fields => #{"foo" => [{"field1", typerefl:integer()}]}
```

File: SCHEMA.md [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+```
+
+##### Virtual struct root
+
+The root struct name exported in the `structs/0` API serves as top level struct's field names.
+like `listener`, `zone` and `broker` in `etc/emqx.conf`.
+
+#### Arrays
+
+Array is a sequence of other types wihch is defined as `{array, Type}`.
```
Comment (k32) [inline SCHEMA.md:92]:
```suggestion
Array is a sequence of other types which is defined as `{array, Type}`.
```

File: SCHEMA.md [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+When starting a Erlang node it requires a system configuration file, (usually named `sys.config`),
+see [Erlang doc](https://erlang.org/doc/man/config.html#sys.config) for more details.
+
+When using HOCON config format, we need a tool to tranfrom a HOCON file to a config file of `sys.config` format.
+`hocon_schema` is such a tool.
+
+### Config mapping
+
+As introduced above, Erlang requires a `sys.config` to bootstrap, the content of this config is essentially
+an Erlang expression which evaluates to an Erlang term (the 'object' in Eralng).
```
Comment (k32) [inline SCHEMA.md:111]:
```suggestion
an Erlang expression which evaluates to an Erlang term (the 'object' in Erlang).
```

File: SCHEMA.md [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+            {Min, Max};
+        _ ->
+            undefined
+    end.
+```
+
+As in the example, a translation callback is provided with the global config,
+specific field values can be retrieved with `hocon_schema:deep_get` API.
+
+### Config integrity vaidation
```
Comment (k32) [inline SCHEMA.md:159]:
```suggestion
### Config integrity validation
```

**Comment by zmstone** [comment]
Thank you for the review @k32 .
I have fixed the typos.

File: SCHEMA.md [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+## Types
+
+There are two high level kinds of data types defined in HOCON schema: `primitve` and `complex`.
+If we think of Erlang data structure as a 'tree', then `primitive` types denote the 'leaves'
+of the 'tree'. aka the terminal values.
+While the `complex` types denote the values which enclose either other `complex` or `primitive` values.
+
+### Primitive types
+
+Most of the `primitive` types are provided by (and also can be extended from) the [typerefl](https://github.com/k32/typerefl) libray.
```
Comment (qzhuyan) [inline SCHEMA.md:15]: s/libray/library

File: SCHEMA.md [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+A union type is in some contexts `one_of` types.
+When data is validated against the schema (recursively), the code enumerates
+the union member types in the defined order until the given data matches any of the union member.
+
+### Config generation
+
+When starting a Erlang node it requires a system configuration file, (usually named `sys.config`),
+see [Erlang doc](https://erlang.org/doc/man/config.html#sys.config) for more details.
+
+When using HOCON config format, we need a tool to tranfrom a HOCON file to a config file of `sys.config` format.
```
Comment (qzhuyan) [inline SCHEMA.md:105]: s/tranfrom/transform

File: SCHEMA.md [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+#### Structs
+
+NOTE: HOCON schema does not support non-struct root level data types. e.g. it is not allowed to
+define a root level schema with just a `integer()` type.
+
+Structs consist of data fields, which can be defined using `hocon_schema` behaviour callbacks.
+
+* `structs/0`: This callback returns all the root level namespaces.
+* `fields/1`: THis callback returns the schema for each data field (in a list, so order matters).
```
Comment (qzhuyan) [inline SCHEMA.md:50]: s/THis/This

File: SCHEMA.md [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+
+When starting a Erlang node it requires a system configuration file, (usually named `sys.config`),
+see [Erlang doc](https://erlang.org/doc/man/config.html#sys.config) for more details.
+
+When using HOCON config format, we need a tool to tranfrom a HOCON file to a config file of `sys.config` format.
+`hocon_schema` is such a tool.
+
+### Config mapping
+
+As introduced above, Erlang requires a `sys.config` to bootstrap, the content of this config is essentially
```
Comment (qzhuyan) [inline SCHEMA.md:110]: I think `sys.config` is optional for erlang.  You may say some application requires 'sys.config' to bootstrap.

File: SCHEMA.md [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+  NOTE: For generic override, see [below](#default_override_rule) for more info
+* `nullable`: set to `true` if this field is allowed to be `undefined`.
+  NOTE: there is no point setting it to `true` if fields has a default value.
+* `sensitive`: set to `true` if this field's value is sensitive so we will obfuscate the log
+  with `********` when logging.
+
+<a name="default_override_rule"></a>
+## Default environment variable override
+
+By default, a field (except for when it's inside an array element) can be overriden by an environment
```
Comment (qzhuyan) [inline SCHEMA.md:205]: s/overriden/overridden

File: SCHEMA.md [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+```
+
+As in the example, a translation callback is provided with the global config,
+specific field values can be retrieved with `hocon_schema:deep_get` API.
+
+### Config integrity validation
+
+Inter-field or even inter-object config validation can be done by implementing
+the `validations` optional callback.
+Validations work similar to translations, only the OK return value is discarded
```
Comment (qzhuyan) [inline SCHEMA.md:163]: OK :: ok | true.

File: SCHEMA.md (hunk 15 above)
Comment (zmstone) [reply SCHEMA.md:110]: I have rephrased it in the above section with a 'usually' word, and removed 'requires ...' here.

**Review by k32 — APPROVED** [review]

**Review by k32 — APPROVED** [review]

**Review by qzhuyan — APPROVED** [review]

**Pull request merged**